jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EventsHandler.java

```diff
@@ -0,0 +1,419 @@
+     * @param status the response status
+     * @param headers the immutable fields of the response object
+     * @see Response#write(boolean, ByteBuffer, Callback)
+     */
+    protected void onResponseBegin(Request request, int status, HttpFields headers)
+    {
+    }
+
+    /**
+     * Invoked before each response content chunk has been written (i.e. from
+     * within the any call to {@link Response#write(boolean, ByteBuffer, Callback)}).
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param last indicating last write
+     * @param content The {@link ByteBuffer} of the response content chunk (readonly).
+     * @see Response#write(boolean, ByteBuffer, Callback)
+     */
+    protected void onResponseWrite(Request request, boolean last, ByteBuffer content)
+    {
+    }
+
+    /**
+     * Invoked after each response content chunk has been written
+     * (i.e. immediately before calling the {@link Callback} passed to
+     * {@link Response#write(boolean, ByteBuffer, Callback)}).
+     * This will always fire <em>before</em> {@link #onResponseTrailersComplete(Request, HttpFields)} is fired.
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param failure if there was a failure to write the given content
+     * @see Response#write(boolean, ByteBuffer, Callback)
+     */
+    protected void onResponseWriteComplete(Request request, Throwable failure)
+    {
+    }
+
+    /**
+     * Invoked after the response trailers have been written <em>and</em> the final {@link #onResponseWriteComplete(Request, Throwable)} event was fired.
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param trailers the written trailers.
+     */
+    protected void onResponseTrailersComplete(Request request, HttpFields trailers)
+    {
+    }
+
+    /**
+     * Invoked when the request <em>and</em> response processing are complete,
+     * just before the request and response will be recycled (i.e. after the
+     * {@link Runnable} return from {@link org.eclipse.jetty.server.HttpChannel#onRequest(MetaData.Request)}
+     * has returned and the {@link Callback} passed to {@link Handler#handle(Request, Response, Callback)}
+     * has been completed).
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param failure if there was a failure to complete
+     */
+    protected void onComplete(Request request, Throwable failure)
+    {
+    }
+
+    private class EventsResponse extends Response.Wrapper
+    {
+        private boolean notifiedOnResponseBegin;
```

Comment:
OK, I know I said I'd say no more... but again naming is confusing... is notified different to fired? No I'll say no more.